[DTLTO][LLD][ELF] Add support for Integrated Distributed ThinLTO

bd1976bris opened the pull request:
This patch introduces support for Integrated Distributed ThinLTO (DTLTO) in ELF LLD.

DTLTO enables the distribution of ThinLTO backend compilations via external distribution systems, such as Incredibuild, during the traditional link step: https://llvm.org/docs/DTLTO.html.

It is expected that users will invoke DTLTO through the compiler driver (e.g., Clang) rather than calling LLD directly. A Clang-side interface for DTLTO will be added in a follow-up patch.

Note: Bitcode members of archives (thin or non-thin) are not currently supported. This will be addressed in a future change.

Testing:
- ELF LLD `lit` test coverage has been added, using a mock distributor to avoid requiring Clang.
- Cross-project `lit` tests cover integration with Clang.

For the design discussion of the DTLTO feature, see: #126654.

bd1976bris committed:
[DTLTO][LLD][ELF] Add support for Integrated Distributed ThinLTO

This patch introduces support for Integrated Distributed ThinLTO
(DTLTO) in ELF LLD.

DTLTO enables the distribution of ThinLTO backend compilations via
external distribution systems, such as Incredibuild, during the
traditional link step: https://llvm.org/docs/DTLTO.html.

It is expected that users will invoke DTLTO through the compiler
driver (e.g., Clang) rather than calling LLD directly. A Clang-side
interface for DTLTO will be added in a follow-up patch.

Note: Bitcode members of non-thin archives are not currently
supported. This will be addressed in a future change.

Testing:
- ELF LLD `lit` test coverage has been added, using a mock distributor
  to avoid requiring Clang.
- Cross-project `lit` tests cover integration with Clang.

For the design discussion of the DTLTO feature, see:
https://github.com/llvm/llvm-project/issues/126654

bd1976bris commented:
Force pushed some minor changes to ensure that all comments made previously on #126654 and #127749 have been addressed here.

bd1976bris commented:
@teresajohnson, I've added some code to this PR (see Input.cpp) to handle ThinLTO bitcode members in thin archives by making the identifier for such members the path to the actual bitcode file on disk for that member. Could a similar approach work to support bitcode members in thin archives for Bazel distribution (--thinlto-index-only)?

It would be great if whatever archive scheme we end up with has a fighting chance of being useful to the Bazel distribution (`--thinlto-index-only`) eventually. The full scheme we have in our downstream codebase supports thin archive bitcode members using the approach in this PR, and supports non-thin archive bitcode members by converting any non-thin archives to thin ones - conceptually - before the linker even looks at the members (the conversion is also cached).

Apologies - I think I meant to raise this question earlier.

bd1976bris committed:
Remove thin-archive bitcode member support

This change removes the DTLTO support for thin-archive bitcode members
from this patch. This makes the patch simpler and more amenable to
review. We are also exploring how to support bitcode members in
non-thin archives. That work is progressing and may require a different
approach for thin-archive support than was originally implemented.

If that is the case, reviewing the current implementation would waste
reviewers' time.

Note that with this support removed, this patch is no longer sufficient
to allow for self-hosting an LLVM build with DTLTO. Theoretically,
--start-lib/--end-lib could be used instead. However, unlike
thin-archives, it's unclear how --start-lib/--end-lib can be easily
used with the LLVM build system.

bd1976bris commented:
> If it helps, I can remove support for thin archive bitcode members in this patch? We’re likely to revisit and revise this area anyway when we add support for non-thin archives and DTLTO. Let me know what you prefer. Thanks!

@maskray and @teresajohnson - I've pushed a change to remove the DTLTO support for thin-archive bitcode members from this patch, and I've updated the PR description accordingly.

This simplifies the patch and makes it more amenable to review. We're also actively exploring how to support bitcode members in non-thin archives, and that work may lead to a different approach for thin-archive support than what was originally proposed here.

If that ends up being the case, reviewing the current implementation would be a potential waste of your time.

Note: without thin-archive support, this patch is no longer sufficient to self-host an LLVM build using DTLTO. While --start-lib/--end-lib could theoretically be used instead, it's not clear (to me) how to integrate that with the current LLVM build system as easily as thin archives.

I hope this change is helpful. Please let me know if you disagree, and I’ll be happy to reinstate.

bd1976bris commented:
Rebased and force-pushed to base this patch on a more recent commit.

MaskRay commented on lld/ELF/Options.td:
```diff
@@ -710,7 +710,17 @@ def thinlto_object_suffix_replace_eq: JJ<"thinlto-object-suffix-replace=">;
-
+def thinlto_distributor_eq: JJ<"thinlto-distributor=">,
+  HelpText<"Distributor to use for ThinLTO backend compilations. If specified, "
+  "ThinLTO backend compilations will be distributed.">;
```
In HelpText, we omit the trailing full stop.

bd1976bris replied:
Thanks. I have added a commit to address this.

MaskRay commented on lld/test/ELF/dtlto/options.test:
```diff
@@ -0,0 +1,41 @@
+# REQUIRES: x86
+
+# Test that DTLTO options are passed correctly to the distributor and
```
`## ` for non-RUN-non-CHECK comments to make them stand out.

MaskRay commented on lld/test/lit.cfg.py:
```diff
@@ -36,6 +36,7 @@
 
 llvm_config.use_default_substitutions()
 llvm_config.use_lld()
+config.substitutions.append(("%llvm_src_root", config.llvm_src_root))
```
I grepped this variable.

clang does it in a different way. Do you know the difference?
```
clang/test/lit.site.cfg.py.in
5:config.llvm_src_root = path(r"@LLVM_SOURCE_DIR@")
```

MaskRay reviewed (commented):
Apologies for the delayed response. I've struggling to manage day work and handle reviews....

MaskRay commented on lld/test/ELF/dtlto/empty.test:
```diff
@@ -0,0 +1,44 @@
+RUN: llc -filetype=obj t1.ll -o t1.o
+RUN: llc -filetype=obj t2.ll -o t2.o
+
+# Check linking succeeds when all input files are ET_RELs.
```
s/ET_RELs/relocatable files.

MaskRay commented on cross-project-tests/dtlto/dtlto.c:
```diff
@@ -0,0 +1,35 @@
+// RUN:   --thinlto-remote-compiler-arg=--save-temps
+
+/// Check that the required output files have been created.
+// RUN: ls | count 10
```
`ls | sort | FileCheck %s` then we can drop DAG and the `count 10` test with `CHECK-NOT: {{.}}`

We can use `sort` in testsuite, `rg 'RUN:.* sort '`

bd1976bris replied:
Nice suggestion - thanks :) I have added a commit to address this.

bd1976bris commented:
> Apologies for the delayed response. I've struggling to manage day work and handle reviews....

No problem! I really appreciate you taking the time to look at this patch.

bd1976bris committed:
Remove trailing full stop from new help text.

bd1976bris committed:
Use "relocatable file" in test comment rather than "ET_REL"

bd1976bris committed:
Improve ls usage/checks in tests.

bd1976bris committed:
"#" -> "##" to make comments stand out more.

bd1976bris committed:
Remove unneeded "-O2" use.

bd1976bris committed:
Remove unneeded "--relocation-model=pic" use.

bd1976bris committed:
Remove unneeded "entry:" lines from textual IR in tests.

bd1976bris committed:
Improve comment in test to explain the presence of the PID in a the file name.

bd1976bris committed:
Remove unneeded linker name from error message check.

MaskRay commented on lld/ELF/InputFiles.cpp:
```diff
@@ -20,6 +20,7 @@
 #include "llvm/ADT/CachedHashString.h"
 #include "llvm/ADT/STLExtras.h"
 #include "llvm/LTO/LTO.h"
+#include "llvm/Object/Archive.h"
```
unneeded?

MaskRay commented on lld/test/ELF/dtlto/index.test:
```diff
@@ -0,0 +1,43 @@
+# REQUIRES: x86
```
This is quite similar to imports.test and can be merged into it. empty.test can be removed if it does not add additional coverage.

bd1976bris replied:
I have pushed a commit to combine the coverage in `imports.test`, `index.test` and `save-temps.test` into one test: `files.test`.

bd1976bris committed:
Removed unnecessary line left over from now removed code.

bd1976bris committed:
Amalgamate imports.test, index.test, and save-temps.test into files.test

bd1976bris committed:
Remove empty.test

bd1976bris committed:
Simplify lld/test/ELF/dtlto/partitions.test

bd1976bris committed:
Move x86 requirement cross-project-test into lit.local.cfg

All cross-project DTLTO tests will use x86 for now.

bd1976bris committed:
Rename the dtlto.c test to ld-dtlto.c since it specifically tests ld.lld

This prepares the test suite for future additions that may cover other
linkers such as lld-link.exe.

bd1976bris commented on lld/test/ELF/dtlto/save-temps.test:
```diff
@@ -0,0 +1,39 @@
+# REQUIRES: x86
```
I have pushed a commit to combine the coverage in `imports.test`, `index.test` and `save-temps.test` into one test: `files.test`.

bd1976bris committed:
Fix missing word in comment sentence

bd1976bris commented:
@MaskRay I’ve pushed commits to address your latest comments. I also made a couple of additional changes that weren’t explicitly requested:

- Reduced the size of partition.test, similar to what was done in the new files.test.
- Made some minor improvements to the cross-project-tests; details are in the relevant commit messages.

Hopefully that’s okay - but let me know if not, and I’m happy to revert.

MaskRay commented:
@bd1976bris The previous comment appears to report failures for the new test dtlto/files.test
```
# RUN: at line 34
z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe @l.rsp
# executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe' @l.rsp
# .---command stderr------------
# | ld.lld: error: cannot open Files\Python310\python.exe: no such file or directory
# `-----------------------------
# error: command failed with exit status: 1
```

MaskRay commented:
> > @bd1976bris The previous comment appears to report failures for the new test dtlto/files.test
> > ```
> > # RUN: at line 34
> > z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe @l.rsp
> > # executed command: 'z:\b\llvm-clang-x86_64-sie-win\build\bin\ld.lld.exe' @l.rsp
> > # .---command stderr------------
> > # | ld.lld: error: cannot open Files\Python310\python.exe: no such file or directory
> > # `-----------------------------
> > # error: command failed with exit status: 1
> > ```
> 
> Thanks! I believe the paths in the response file need to be quoted to handle spaces - e.g., "C:\Program Files\Python310\python.exe". Would you prefer I revert this PR and open a new one with the fix, or should I just push a point fix to address the buildbot failure?

I’d prefer that you fix it forward.... This bot appears to be managed by your team.

bd1976bris commented:
> I’d prefer that you fix it forward.... This bot appears to be managed by your team.

Thanks: https://github.com/llvm/llvm-project/pull/146749.

> ... This bot appears to be managed by your team.

Yes — embarrassing! I believe this is a purely upstream bot that isn't used for any of our downstream testing. It looks like Python was installed in a different location compared to the other test machines.
